accepted/0000-remove-dependency-on-fulcio.md

```diff
@@ -0,0 +1,69 @@
+# RFC for linking packages to their source and build without using Sigstore Fulcio
+
+## Summary
+
+As it is showed in a recent paper that Sigstore Fulcio is an unnecessary trusted third party for linking npm packages to their source and build, we suggest the npm team consider the new approach presented in that paper which can be found at:
+
+https://eprint.iacr.org/2023/003
+
+The new approach not only eliminates the unnecessary dependency on an external black-box CA but also accelerates the signing process.
+
+## Motivation
+
+Last year an RFC for linking npm packages to their source and build using Sigstore Fulcio was proposed and accepted:
+
+https://github.com/npm/rfcs/pull/626
+
+However, the paper mentioned above shows the same result can be achieved without using Sigstore Fulcio at all. Since relying on an unnecessary black-box CA poses a serious threat to the npm community, we draft this RFC.
+
+## Detailed Explanation
+
+The key observation made by the paper mentioned above is the following:
+
+- Instead of setting the Audience claim of a requested OIDC token to be `sigstore` (as proposed by the previous RFC), a CI/CD workflow can in fact set the Audience claim to be something like `sigstore?pk=PubKey`, where `PubKey` is (the Base64 encoding of) the public key used by the CI/CD workflow for signing the software artifact.
+
+In other words, such an OIDC token can be understood as a witness signed by the CI/CD service, with the Audience claim encoding the workflow's intention to use the public key to represent it. Thus, instead of exchanging an OIDC token (and a public key) for a Fulcio certificate, we can directly use the OIDC token in place of the certificate (which is, after all, transcribed from the OIDC token).
+
+As for verification, instead of hardcoding the root certificate of Fulcio, we can just hardcode the public key of the corresponding CI/CD service (aka OIDC provider), which is part of the root of trust of Sigstore Fulcio (so that no new trust is introduced).
+
+## Rationale and Alternatives
+
+- The previous RFC is an alternative that depends on an external black-box CA
+- This RFC requires no external black-box CA
+- The paper mentioned above suggests Sigstore Rekor, also proposed in the previous RFC, can be directly replaced by standard RFC 3161 Time Stamping Authority (TSA) servers hosted by established CAs such as DigiCert and Sectigo for business grade SLAs
```

Comment:
Rekor, a transparency log, is not equivalent to a timestamping authority. Rekor currently provides two things for each entry:
* The time the entry was logged in Rekor
* A way to cryptographically prove inclusion

The benefit of Rekor is that all entries are logged publicly in an append-only log. The second point here lets you prove that an entry is in the log, which means that the artifact and signature is publicly auditable. As a verifier, you shouldn't trust an entry that isn't publicly logged, which forces malicious actors into the public and can therefore be detected.

The first point is just timestamping, which isn't new to code signing. See Authenticode, Microsoft's approach to code signing that's been around for decades. You could use a timestamping authority to provide a timestamp instead of Rekor - We actually just added support for this in Cosign, Sigstore's CLI.

Using a TSA in addition to Rekor gives you both a third-party to assert time, and auditability. Using a TSA without Rekor is just timestamping, and is no different than previous approaches to code signing that lack transparency.

Reply:
Unfortunately, to verify if a signature is on Rekor, you need to either check it online (which renders a signature less meaningful as one can just check the hash listed in a GitHub release note online) or trust a Rekor-signed proof offline (which raises many questions such as why a proof that never expires should be trusted, why the proof-signing key can never be compromised, what if no one will do online check when everyone is used to the offline experience, and et al). However, since this RFC is about Fulcio, I am not against using Rekor for the time being.

Reply:
> …you need to either check it online (which renders a signature less meaningful as one can just check the hash listed in a GitHub release note online)

A GitHub release note isn’t immutable. An attacker could compromise a repository and change the artifact hash to point to one they control

> or trust a Rekor-signed proof offline (which raises many questions such as why a proof that never expires should be trusted, 

Why should a proof expire? By asserting that a code signing certificate was valid at a time when either the entry was added to a log or a signed time stamp was generated, a verifier can know that a signer was in possession of the artifact signing key. A signature doesn’t need to expire, and if it did, it would be a significant burden on maintainers and verifiers to have to re-sign or fetch updated signatures. 

> why the proof-signing key can never be compromised

It could be compromised, and this is the risk of offline proofs not being easy to verify. Note that in certificate transparency, a verifier should verify an offline proof in an online manner at some point, but this mechanism hasn’t been implemented at scale.
Using TUF, we can respond quickly to compromises by shipping new key material with the TUF root. 

> what if no one will do online check when everyone is used to the offline experience, and et al)

that’s up to a verifier. Online verification will always provide stronger assurances.